### ISSUE TITLE
Enable stc to write jars to a local directory instead of just publishing them

### ISSUE THREAD

**Issue by jvican** [ScalablyTyped/Converter#480]
I'd like to suggest that in addition to having the option to publish a package locally and reuse it like we would do if the dependency was available in a remote repository, we have the ability to write jars to a local folder.

The benefit of this approach is the following:

- Reproducibility and no need for downloading type dependencies
- We have a stable snapshot of all type dependencies in a folder
- We remove the dependency on `stc` and `publishLocal` during CI

Users that opt in for this mode should always rerun `stc` when updating their `package.json`, and they even have the ability of creating an npm script that will do that automatically, which is nice as the process of using js dependencies becomes easier.

There's built-in support in Scala build tools as well as scala-cli to import just a custom jar

**Comment by oyvindberg** [comment]
Yeah, that's not a bad idea! You mean using the sbt mechanism for unmanaged jar files, typically inside `lib` folders and so on? and then checking in the jar files?

I've seen some libraries (aws-sdk, googleapis, material-ui) end up on the wrong side of 100 megabytes, but it's quite rare.

**Comment by oyvindberg** [comment]
Well `std` is ~18 megabytes, and a dependency of all generated jars. so every git repo would get that. and it's updated every once in a while. I guess not everybody will be happy about that.

**Comment by jvican** [comment]
It's their choice 😄 I would gladly take fat jars in exchange for the above properties.

If you really want to optimize this, then we can look at why `stc` takes up 18mb, but I don't think it'll be an issue. Ideally, they would be slim, but this is just another trade-off.

**Comment by jvican** [comment]
Huh, it's funny how whenever I get the need to use ScalablyTyped I just go back to this thought. I just wanted to revisit this, and I actually follow up on something that I didn't get last time around. Why does the generated code need the std, which is 18 megabytes?

**Comment by oyvindberg** [comment]
> Huh, it's funny how whenever I get the need to use ScalablyTyped I just go back to this thought. I just wanted to revisit this, and I actually follow up on something that I didn't get last time around. Why does the generated code need the std, which is 18 megabytes?

It's a dependency, downstream repositories will refer to types in it. Scalac will complain if it's not there.

Technically it's possible to cut the dependency, with some loss of fidelity. For instance it would be easy enough to write a transformation which rewrites all references to types in `std` to corresponding types in scala-js-dom (already in place), or rewrite to `Any` otherwise (this would be new). 

---

So anyway, to shine some light on the solutions *currently* in place, I tried to detail the different ways I see of using ST for writing applications. Scenario 1 is likely a good solution for you. I'll iterate a bit more on the text and publish it to scalablytyped.org below.

## Solutions to slow build times, reproducible builds for *applications*. 

If you're developing a library you should in all cases use library developer mode and publish it to a maven repository.

### Scenario 1 - Cache ST `run files` in CI

When sbt runs the `stImport` task to run the conversion, it calculates `ImportTypings.Input` from your build file, and digests it:
<details>
  <summary>Example json input</summary>
<pre><code>
  {
    "converterVersion": "1.0.0-beta39+53-317b19f3-SNAPSHOT",
    "conversion": {
      "useScalaJsDomTypes": true,
      "flavour": "slinky",
      "outputPackage": "typings",
      "enableScalaJsDefined": {
        "NoneExcept": {
          "values": []
        }
      },
      "stdLibs": [
        "dom",
        "es6"
      ],
      "expandTypeMappings": {
        "AllExcept": {
          "values": [
            "prop-types",
            "react",
            "std"
          ]
        }
      },
      "ignored": [
        "typescript"
      ],
      "versions": {
        "scala": "2.13.2",
        "scalaJs": "1.10.0"
      },
      "organization": "org.scalablytyped",
      "enableReactTreeShaking": {
        "NoneExcept": {
          "values": []
        }
      },
      "enableLongApplyMethod": false,
      "privateWithin": null,
      "useDeprecatedModuleNames": false
    },
    "wantedLibs": {
      "csstype": "2.6.11",
      "downshift": "6.0.5",
      "prop-types": "15.7.3",
      "react": "16.9.42",
      "react-dom": "16.9.8"
    }
  }
</code></pre>

</details>

This digests to `050c397b8c0c5704600ee81db06bbf21`, and is found in `$HOME/Library/Caches/ScalablyTyped/runs/050c397b8c0c5704600ee81db06bbf21.json` on my computer. Contents, digests and paths will obviously vary.

In that file is also stored the output:
<details>
  <summary>Example json output</summary>
<pre><code>
"allProjects": [
      {
        "Mangled": {
          "mangledArtifact": "csstype_sjs1_2.13",
          "dep": {
            "ScalaJs": {
              "org": "org.scalablytyped",
              "name": "csstype",
              "version": "2.6.11-0ac64f"
            }
          }
        }
      },
      {
        "Mangled": {
          "mangledArtifact": "prop-types_sjs1_2.13",
          "dep": {
            "ScalaJs": {
              "org": "org.scalablytyped",
              "name": "prop-types",
              "version": "15.7.3-4e9ecf"
            }
          }
        }
      },
      {
        "Mangled": {
          "mangledArtifact": "react_sjs1_2.13",
          "dep": {
            "ScalaJs": {
              "org": "org.scalablytyped",
              "name": "react",
              "version": "16.9.42-5cfe08"
            }
          }
        }
      },
      {
        "Mangled": {
          "mangledArtifact": "std_sjs1_2.13",
          "dep": {
            "ScalaJs": {
              "org": "org.scalablytyped",
              "name": "std",
              "version": "3.9-cb5788"
            }
          }
        }
      },
      {
        "Mangled": {
          "mangledArtifact": "react-dom_sjs1_2.13",
          "dep": {
            "ScalaJs": {
              "org": "org.scalablytyped",
              "name": "react-dom",
              "version": "16.9.8-e00964"
            }
          }
        }
      },
      {
        "Mangled": {
          "mangledArtifact": "downshift_sjs1_2.13",
          "dep": {
            "ScalaJs": {
              "org": "org.scalablytyped",
              "name": "downshift",
              "version": "6.0.5-60cafe"
            }
          }
        }
      ]
</code></pre>

</details>

That way we can answer instantly if a conversion needs to run. So this is all static information, and can be cached, so that's exactly what you can do on CI. If you make sure that 
`~/.ivy2/local` and `~/.cache/ScalablyTyped/runs` is cached in CI, it'll just run the conversion when needed. See for instance [circleci config for a demo repo](https://github.com/ScalablyTyped/SlinkyDemos/blob/master/.circleci/config.yml#L41-L42) (note that it caches too much, the ST path should have `/runs` appended optimally)

| Requirement | Score |
| - | - |
| fast? skips conversion in typical build? | 👍 unless on changes or fresh cache |
| automatically updated on newer versions of npm libraries or ST config? | 👍|
| all builds use the exact same jars | 👎 built for each computer |
| ST plugin necessary in application build| yes|
| Infrastructure overhead| little, need to tweak CI script|

### Scenario 2 - Remote Cache

The run files is also the basis for the [Remote Cache feature](https://scalablytyped.org/docs/remotecache). 
It's a bit more reliable, but requires a central place to store jar and json files

| Requirement | Score |
| - | - |
| fast? skips conversion in typical build? | 👍 conversion runs when *you* change something, and then you manually push |
| automatically updated on newer versions of npm libraries or ST config? | 👍, but manual push so other won't recompute jars|
| all builds use the exact same jars | 👎 yes, but with a race condition.  |
| ST plugin necessary in application build| yes|
| Infrastructure overhead| more, needs a central storage location like S3|


### Scenario 3 - Publish jar files to maven outside your build

You'll typically have a separate project or script to generate typing libraries.
This approach is more flexible than just caching, because you can for instance change the generated code before publish.
This should be the default approach for sharing typing libraries between different codebases.

| Requirement | Score |
| - | - |
| fast? skips conversion in typical build? | 👍  |
| automatically updated on newer versions of npm libraries or ST config? | 👎 |
| all builds use the exact same jars | 👍  |
| ST plugin necessary in application build| no |
| Infrastructure overhead| needs a maven repository and a build/script to publish typing libraries|

### Scenario 4 - [library developer mode](https://scalablytyped.org/docs/library-developer) in *automatic* mode 

This used to be how the earliest versions of the ST sbt plugin worked. It was replaced by the current ST application plugin to solve compile times primarily. 

The viability of this being a success in your app likely depends on setting up `stMinimize` so you massively decrease the amount of generated code you have to compile. Using it correctly typically means that `std` is reduced to a handful of files

| Requirement | Score |
| - | - |
| fast? skips conversion in typical build? | 👍 skips conversion, but incurs more compiling  |
| automatically updated on newer versions of npm libraries or ST config? | 👍  |
| all builds use the exact same jars | 👎 source code. ST tries to be deterministic, but may have bugs  |
| ST plugin necessary in application build| yes |
| Infrastructure overhead| a bit more setup in build|


### Scenario 5 - [library developer mode](https://scalablytyped.org/docs/library-developer) in *manual* mode 

In this mode, it is envisioned that you check in the generated source code. As such you get it under version control, can track changes over time. This is great because it brings consistent builds and clarity.

| Requirement | Score |
| - | - |
| fast? skips conversion in typical build? | 👍 skips conversion, but incurs more compiling  |
| automatically updated on newer versions of npm libraries or ST config? | 👎 , need to run `stImport` manually and then check in result  |
| all builds use the exact same jars | 👍 same source code |
| ST plugin necessary in application build| yes |
| Infrastructure overhead| a bit more setup in build|

**Comment by dispalt** [comment]
Are there instructions @oyvindberg on how to do Scenario 3, I seem to remember seeing something but I can't seem to find it.
